### PR TITLE
[ADD] web_widget_slick: Add module

### PR DESCRIPTION
**Pull request by hughesbm**
## Slick Carousel Widget

The `web_widget_slick` module provides a [Slick](http://kenwheeler.github.io/slick/) Carousel widget for use in Odoo.

The `web_widget_slick_example` module provides an example implementation; views can be found in Settings / Slick

### Usage

Default usage is on a One2many attachment field, as defined below:

    class SlickExample(models.Model):
        _name = 'slick.example'
        _description = 'Slick Example Model'
        image_ids = fields.One2many(
            name='Images',
            comodel_name='ir.attachment',
            inverse_name='res_id',
        )

Assuming the above model, you would add a Slick Carousel on the `image_ids` column by using the following field definition in the model's form view:

    <field name="image_ids" widget="one2many_slick_images" options="{}"/>

### Options

The widget passes options directly through to Slick, [so any setting available to Slick](http://kenwheeler.github.io/slick/#settings) is available to the widget.

Additionally, Odoo-specific options `fieldName` and `modelName` are available; see readme for details.

### CODE REVIEW

File: web_widget_slick/README.rst
Comment (lasley) [inline web_widget_slick/README.rst]: "for use in the Odoo backend web interface"

File: web_widget_slick/static/src/css/slick.css
Comment (lasley) [inline web_widget_slick/static/src/css/slick.css]: Change to LESS

I don't think you'll need to do anything other than the file extension and folder for the most part though.

Also this was like 2015 or something when I made this - double check the timeline for copyright.

File: web_widget_slick/static/src/css/slick.css
Comment (lasley) [inline web_widget_slick/static/src/css/slick.css]: Ok so I never really liked this part - the arrows look like crap if I remember right. We should maybe try to use fontawesome or glyphs, and maybe the [built-in LESS colors](https://github.com/odoo/odoo/blob/10.0/addons/web/static/src/less/variables.less)

**Review by lasley — CHANGES_REQUESTED** [review]
Few inline comments. High level notes:

* Only pull in the library files we need, so in this case just the slick JS (make sure the version number is in header)
  * I don't think we use the theme, but if you want to give it a shot go for it but use the LESS version. It's probably worth trying - it probably looks nice
* Switch the icon to OCA

**Comment by hughesbm** [comment]
Styling has been updated to be more Odoo-like. Arrow buttons now use left/right font awesome chevrons (also used in paging buttons in Odoo) and LESS mixins of button classes to match other brand-colored buttons.

screen shot 2017-08-24 at 1 32 26 pm

**Comment by lasley** [comment]
Now that I'm thinking about it, let's go ahead and move the example module in too. It sucks, but I can't find any model in core that has a o2m attachments field that we could use 😦

**Comment by hughesbm** [comment]
The example module has been added. 👍  ayyyyyy 👍

File: web_widget_slick/README.rst
Comment (lasley) [inline web_widget_slick/README.rst]: Add a note here about functionally testing this module using the example

**Comment by hughesbm** [comment]
PhantomJS is failing to find the field template on [Travis](https://travis-ci.org/OCA/web/jobs/268572188#L905), but works without any problems locally. The tests themselves all pass when run in the browser. Functional tests on runbot work as expected, too.

Any ideas on what could cause this?

File: web_widget_slick/README.rst
Comment (lasley) [inline web_widget_slick/README.rst:51]: How does a functional reviewer find the menu for testing? Assume that a functional reviewer does not review code, so pointing them to code is no bueno

**Comment by pedrobaeza** [comment]
Testing a bit this widget, I see that dragging over an image, there's a strange behavior that doesn't happen if you drag in the carousel outside an image. I see also that this control is readonly. Do you plan to add features for allowing to add or delete images?

**Comment by lasley** [comment]
> I see also that this control is readonly. Do you plan to add features for allowing to add or delete images?

Not at this time, but that would be a good thing for roadmap.

We are, however, going to be implementing image editing with a module combining this with `web_widget_darkroom` from #595. The slick part will serve as an image browser with the darkroom part as the editor.

**Comment by hughesbm** [comment]
@pedrobaeza I believe I found and dealt with the dragging behavior you mentioned; if it's not resolved for you and I just worked on a different dragging issue instead, please let me know! 😄 

@lasley I updated the readme with more detailed instructions, and added a roadmap item regarding adding/deleting images from the carousel.

I also fixed some other issues related to lazy-loaded images not displaying properly while being sized responsively, especially when using the Slick grid settings.

**Comment by pedrobaeza** [comment]
@hughesbm indeed, the dragging issue I mentioned is fixed with this commit, thanks!

**Comment by hughesbm** [comment]
Update: I was able to reproduce the PhantomJS failure to find the widget template locally, as well, so it's not a Travis issue. The template is loaded fine on subsequent test runs, but can't be found on the initial run.

**Comment by hughesbm** [comment]
The testing issue has been fixed! 🎉

File: web_widget_slick/static/src/js/web_widget_slick.js
Comment (lasley) [inline web_widget_slick/static/src/js/web_widget_slick.js]: Move this into a method `slickRemove`, delete the `var self = this;` above, and use proxy to send to destroy:

```
$imgs.each($.proxy(this.slickRemove, this));
```

File: web_widget_slick/static/src/js/web_widget_slick.js
Comment (lasley) [inline web_widget_slick/static/src/js/web_widget_slick.js]: Use `this`

File: web_widget_slick/static/src/js/web_widget_slick.js
Comment (lasley) [inline web_widget_slick/static/src/js/web_widget_slick.js]: Move this stuff into a `slickRender` then proxy to it just like recommended in `slickRemove`

In JS if you find yourself using the ol' `this=self` trick, you're probably nesting too much. I know I probably did this, so not your fault, but good to know regardless.

File: web_widget_slick/static/src/js/web_widget_slick.js
Comment (lasley) [inline web_widget_slick/static/src/js/web_widget_slick.js]: Can just call this `web_widget_slick`

File: web_widget_slick/static/src/xml/web_widget_slick.xml
Comment (lasley) [inline web_widget_slick/static/src/xml/web_widget_slick.xml]: `t-attf-class="o_form_field {{ widget.widget_class }}"`

File: web_widget_slick/static/tests/js/web_widget_slick.js
Comment (lasley) [inline web_widget_slick/static/tests/js/web_widget_slick.js]: I typically suffix my tours with `.tour`, but that's stylistic so feel free to leave this.

File: web_widget_slick/static/tests/js/web_widget_slick.js
Comment (lasley) [inline web_widget_slick/static/tests/js/web_widget_slick.js]: Hrmmmm this tabbing is bugging the crap out of me. I think it would be best to just leave the function declaration on the first line - it won't be that much longer than your other ones and we don't have a 79 char hard limit like we do in Python.

**Review by lasley — CHANGES_REQUESTED** [review]
Thanks @hughesbm - comments inline

File: web_widget_slick/static/tests/js/web_widget_slick.js
Comment (hughesbm) [reply web_widget_slick/static/tests/js/web_widget_slick.js]: These are just unit tests, not a tour. 😄

**Review by lasley — APPROVED** [review]
Thanks @hughesbm

**Comment by hughesbm** [comment]
I just noticed there are display problems with multiple widgets per page, so this module still needs work.

**Comment by hughesbm** [comment]
The latest commit fixes display issues related to use of `<group>` and `<sheet>` tags. The widget should now display correctly and be sized appropriately.

File: web_widget_slick_example/views/slick_example_view.xml
Comment (lasley) [inline web_widget_slick_example/views/slick_example_view.xml]: Put this in a group to prove your fix works please

**Review by lasley — APPROVED** [review]
Great job @hughesbm

**Review by felixvillafranca — APPROVED** [review]
Great¡

**Comment by pedrobaeza** [comment]
@lasley squash at your taste before merging.

**Comment by lasley** [comment]
I ended up reordering some of the commits as well so that we could have a clean 4 commits - one for each module creation by me, then another for each module upgrade by Brent.

I'll merge once we get a CI 🍏

**Pull request merged**
